flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/ProcTimeInnerJoin.scala

```diff
@@ -298,16 +289,17 @@ class ProcTimeInnerJoin(
     while (keyIter.hasNext && nextTimer == 0) {
       val curTime = keyIter.next
       if (curTime < expiredTime) {
-        expiredList.add(curTime)
+        listToRemove.add(curTime)
       } else {
         nextTimer = curTime
       }
     }
 
-    var i = 0
-    while (i < expiredList.size) {
-      rowMapState.remove(expiredList.get(i))
-      i += 1
+    var i = listToRemove.size() - 1
+    while (i >= 0) {
+      rowMapState.remove(listToRemove.get(i))
+      listToRemove.remove(i)
```

Comment:
I think it's even more efficient to traverse the `listToRemove` without modifying it and calling `listToRemove.clear()` afterwards.